dev/update_requirements.py

```diff
@@ -0,0 +1,129 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility script for updating examples' "requirements.txt".
+
+WARNING: Updates examples/*/requirements.txt in place, so you probably want to
+run this in a clean working directory...
+
+This script is useful to automatically update the requirements of the individual
+examples. It's desirable to have these dependencies fully specified and
+guaranteed to be complete in order to download an example directly from Github
+and run it on a virtual machine without any additional manual package
+installation (note though that jax/jaxlib must already be installed).
+
+To update all "requirements.txt" the program is run with a `--versions` argument
+that specifies the version of every package that is required by any of the
+examples. If any of the versions is not specified, the program will fail with an
+error.
+
+Example invocations (from the Flax repo root directory):
+
+python dev --versions='absl-py-0.12.0 clu-0.0.6 flax-0.3.6'
+
+The actual pypi-version list can be read copied from the "Install Dependencies"
+step from the latest Github build action:
+https://github.com/google/flax/actions/workflows/build.yml
+
+Alternatively, the list can also be provided from the local environment with:
+
+python dev --versions="$(pip freeze | sed s/==/-/g) flax-0.3.6"
+"""
+
+import pathlib
+import re
+import sys
+
+from absl import app
+from absl import flags
+
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string(
+    'versions',
+    None,
+    'space-separated list of pkg-name-1.0.0 pairs. can be obtained from local '
+    'environment with '
+    '`--version="$(pip freeze | sed s/==/-/g) flax-0.3.6"` '
+    '(note the flax version "override") '
+    'or from the "install dependencies" step in the github build action '
+    'https://github.com/google/flax/actions/workflows/build.yml')
+flags.DEFINE_bool('verbose', False, 'enables verbose output.')
+flags.DEFINE_list('ignore', ['jax'], 'packages not to add to requirements.')
+
+
+import_re = re.compile(r'(?:from|import)\s+(\w+)')
+# maps `import cv2` to `pip install opencv-python`
+pkg_map = {
+  'absl': 'absl-py',
+  'atari_py': 'atari-py',
+  'cv2': 'opencv-python',
+  'ml_collections': 'ml-collections',
+  'PIL': 'Pillow',
+  'tensorflow_datasets': 'tensorflow-datasets',
+  'tensorflow_text': 'tensorflow-text',
+}
+standard_libs = set('codecs collections dataclasses datetime enum functools math multiprocessing itertools os pathlib random re sys tempfile time typing unicodedata warnings'.split(' '))
+
+
+def main(argv):
+  del argv
+  assert FLAGS.versions, 'Must specify --versions.'
```

Comment:
You can also do `flags.mark_flag_as_required('versions')` after you defined the flag.